### DISCUSSION TITLE
Dynamically Rendering Imagick Image in Wizard Form - "Undefined variable $imageUrl" Error

### DISCUSSION THREAD

**Discussion by nccirtu** [filamentphp/filament#14251] [Help]
### Package

Form builder

### Package Version

V3

### How can we help you?

I'm developing an application using Laravel 10 and Filament V3. In my application, I have a wizard form where I want to dynamically generate an image using Imagick based on the selected color and font. The image should be rendered immediately after the selection, not after the form is submitted.

However, I keep encountering the error "Undefined variable $imageUrl" when the page loads. I suspect there might be an issue with how I'm generating or passing the image.

Here’s my code so far:

`Forms\Components\Wizard\Step::make('Nummern')
                        ->icon('heroicon-m-numbered-list')
                        ->schema([
                            Forms\Components\Section::make([
                                Forms\Components\Split::make([
                                    // Auswahl der Schriftart
                                    Forms\Components\Section::make('Auswahl der Schriftart für Nummern')
                                        ->description('Wähle eine passende Schriftart für den Druck aus')
                                        ->schema([
                                            Select::make('font_type_player_numbers')
                                                ->label('Auswahl der Schriftart für Schriftzüge')
                                                ->options([
                                                    'AcademyFilled3D.ttf' => 'Academy 3D',
                                                    'Libero.ttf' => 'Libero',
                                                    'Prisma.ttf' => 'Prisma',
                                                ])
                                                ->live()
                                                ->reactive()
                                                ->afterStateUpdated(function ($state, $get, $set) {
                                                    $set('fontType', $state); // Synchronisiere die Schriftart
                                                }),

                                            // Farbauswahl
                                            TextInput::make('font_color_player_numbers')
                                                ->label('Schriftfarbe')
                                                ->type('color')
                                                ->default('#000000')
                                                ->live()
                                                ->reactive()
                                                ->afterStateUpdated(function ($state, $get, $set) {
                                                    $set('fontColor', $state); // Synchronisiere die Farbe
                                                }),
                                        ]),

                                    // Vorschau Sektion
                                    Forms\Components\Section::make('Vorschau')
                                        ->description('Hier eine Vorschau')
                                        ->schema([
                                            // Hier binden wir die Vorschau an die Livewire-Komponente
                                            ViewField::make('vorschau')
                                                ->view('livewire.text-image-component'),
                                        ])->columns(2),
                                ]),
                            ]),
                        ]),`

//Livewire Component 

`class TextImageComponent extends Component
{
    public $fontType = 'AcademyFilled3D.ttf'; // Standard-Schriftart
    public $fontColor = '#000000'; // Standard-Schriftfarbe
    public $text = 'Vorschau'; // Text für die Vorschau
    public $imageUrl = ''; // Bild-URL

    public function mount(): void
    {
        // Initialisiere die Vorschau
        $this->updatePreview(); // Erstelle das Bild beim Mount
    }

    public function render(): View|Factory|Application
    {
        return view('livewire.text-image-component', [
            'imageUrl' => $this->imageUrl, // Übergibt das generierte Bild an die View
        ]);
    }

    public function updatePreview(): void
    {
        $imagePath = $this->createImage($this->text, $this->fontType, $this->fontColor);
        $this->imageUrl = Storage::url($imagePath);

        // Debugging: Zeige die generierte URL an
        dd($this->imageUrl);
    }

    public function createImage($text, $font, $color): string
    {
        $imageName = 'text_image_' . time() . '.png';
        $imagePath = storage_path('app/public/' . $imageName);

        $imagick = new Imagick();
        $draw = new ImagickDraw();
        $draw->setFillColor(new ImagickPixel($color));
        $draw->setFont(public_path('fonts/' . $font));
        $draw->setFontSize(48);

        $imagick->newImage(800, 200, new ImagickPixel('#ffffff'));
        $imagick->setImageFormat('png');
        $imagick->annotateImage($draw, 50, 100, 0, $text);
        $imagick->writeImage($imagePath);

        $imagick->clear();
        $imagick->destroy();

        return 'public/' . $imageName;
    }
}`

//View

`<div>
    @if ($imageUrl)
        <img src="{{ $imageUrl }}" alt="Vorschau Bild" />
    @else
        <p>Keine Vorschau verfügbar</p>
    @endif
</div>`

I expected the image to be dynamically generated and displayed in the preview section as soon as the user selects a color and font. To achieve this, I tried binding the preview section to the Livewire component and using live() and reactive() methods on the form fields to trigger the image generation in real-time.

However, instead of showing the generated image, I kept getting the "Undefined variable $imageUrl" error. I have also tried debugging by dumping the $imageUrl value, which showed the correct URL but still caused the error in the view.

**Comment by leandrocfe** [comment]
The code format is quite poor here. Please format it properly and provide only the relevant code.

**Reply by nccirtu** [reply]
Thank you for the feedback. I am currently trying to create a custom field that generates a preview image based on the selected values for font color and font type. Initially, I tried using a Livewire component, but that didn’t work as expected. In my current custom field, the update method doesn’t seem to trigger, so the preview is not being updated. Here is the code for the custom field and the form. I would greatly appreciate any feedback or suggestions.

`
//My Form
Forms\Components\Wizard\Step::make('Nummern')
                        ->icon('heroicon-m-numbered-list')
                        ->schema([
                            Forms\Components\Section::make([
                                Forms\Components\Split::make([
                                    // Auswahl der Schriftart
                                    Forms\Components\Section::make('Auswahl der Schriftart für Nummern')
                                        ->description('Wähle eine passende Schriftart für den Druck aus')
                                        ->schema([
                                            Select::make('fontType')
                                                ->label('Auswahl der Schriftart für Schriftzüge')
                                                ->options([
                                                    'AcademyFilled3D.ttf' => 'Academy 3D',
                                                    'Libero.ttf' => 'Libero',
                                                    'Prisma.ttf' => 'Prisma',
                                                ])
                                                ->reactive()
                                                ->live(),

                                            // Farbauswahl
                                            TextInput::make('font_color_player_numbers')
                                                ->label('Schriftfarbe')
                                                ->type('color')
                                                ->default('#000000')
                                                ->reactive()
                                                ->live(), // Direkt auf Änderungen reagiere
                                        ]),

                                    // Vorschau Sektion
                                    Forms\Components\Section::make('Vorschau')
                                        ->description('Hier eine Vorschau')
                                        ->schema(components: [
                                            ClubNamePreview::make('imageUrl')
                                                ->label('Text Vorschau Bild')
                                                ->reactive()
                                                ->live(),
                                        ])->columns(2),
                                ]),
                            ]),
                        ]),

//My Custom Field
class ClubNamePreview extends Field
{
    protected string $view = 'forms.components.clubNamePreview';

    public $clubName = 'Vereinsname';
    public $fontFamily = 'Roboto';
    public $fontColor = '#000000';

    public function setClubName($name): static
    {
        $this->clubName = $name;
        return $this;
    }

    public function setFontFamily($fontFamily): static
    {
        $this->fontFamily = $fontFamily;
        return $this;
    }

    public function setFontColor($fontColor): static
    {
        $this->fontColor = $fontColor;
        return $this;
    }

    // Hier geben wir die Daten explizit an das Blade-Template weiter
    protected function getViewData(): array
    {
        return [
            'clubName' => $this->clubName,
            'fontFamily' => $this->fontFamily,
            'fontColor' => $this->fontColor,
        ];
    }
}
`
![Bild Step](https://github.com/user-attachments/assets/fdde5b4a-dfe3-4305-aab5-40d84fba08da)
